[BUGFIX] Fix el fn query argument type and update corresponding return type

katywings opened the pull request:
This is how typescript resolves the types on this pr:
![redomElTypes](https://user-images.githubusercontent.com/4012401/60052123-ce4e9380-96d4-11e9-8590-3b364683d466.png)

~The only one case not having a nice element resolution is the `aEl`. Reason is, because the this.el definition in the function cannot be recognized by typescript, I previously decided to just fallback to HTMLElement in this case, but I think we should actually change this case to RedomComponent.~

Edit: we removed Function as a valid query argument type so the aEl case has to be typed by the consumer manually.

---

Used code for type testing:
```typescript
import { el } from '.'

class TdComponent {
    el = el('td')
}

const els = {
    tableEl: el(document.createElement('table')),
    tdComponent: el(TdComponent),
    divEl: el('div'),
    aEl: el(function() {
        this.el = el('a')
    }),
    buttonEl: el(function() {
        return {
            el: el('button')
        } 
    }),
    inputEl: el('input')
}
```

pakastin commented:
Works still in progress @katywings?

katywings commented:
@pakastin: @raulil and I discussed it months ago for a while in Slack, the slack archives dont go back so far but I think he didnt like this line:
https://github.com/lufrai/redom/blob/bb805774d5940f06c864634c48686116a6038c95/index.d.ts#L110

Hes point (as I remember) was that "Function" keyword shouldnt be used in types), but I think he didnt mention any alternative.

So we discussed around the philosophical question if types in general should strictly disallow use cases that are actual valid javascript when those use cases cannot be 100% typed.

We werent able to solve the discussion. 

And I couldnt just close this pr for philosophical reasons...

So the state is: one valid js usecase cannot be typed in a 100% strict way. For strictness sake we can just disallow this usecase completely or we live with not 100% strictness.

If you (@pakastin) dont like pragmatic strictness then this pr can be closed. If you are happy with only 99.9% strict types then @raulil should say if the pr is otherwise ok.
If it is I would have to rebase it first again though :D

RauliL commented:
I still think that introducing `Function` here would reduce the quality of type definitions, because it's so vague that any function would pass as valid RE:DOM component. I'm myself fan of strict typing and think that the use case shown in the screenshot is a very rare case even if it works with JavaScript and is kinda bad convention anyways.

katywings commented:
Maybe the consumer solution could be written like this:

```typescript
{
aEl: el(function() {
        this.el = el('a')
} as RedomComponent)
}
```

I have to try it out when I am on a pc. If it works we could leave the `Function` condition away in the types - with the drawback that this solution will not work for ts-check js consumers as far as I know.

@raulil would you prefer this consumer solution? Anything else from your side regarding the pr?

RauliL commented:
Regarding the provided example, I think the solution would be either to use
```TypeScript
aEl: el('a')
```
or
```TypeScript
aEl: { el: el('a') }
```
or if you want to construct it somehow dynamically
```TypeScript
aEl: (() => {
  const result = { el: el('a') };
  ...
  return result;
})()
```

katywings commented:
@raulil yeah those work too in this case, but not in the list(...) case where you have to specify a class / function and then you have to use the `as RedomComponent` as consumer. Anyways, I slowly loose motivation to further spend time on this pr only to discuss without any progress... 

Anything else from your side regarding the pr? Mergeable after removing function and rebase?

RauliL commented:
Would it work if you change `Function` to `() => { el: HTMLElement }`?

katywings commented:
@raulil the function-return case already exists on this line :) :
https://github.com/lufrai/redom/blob/bb805774d5940f06c864634c48686116a6038c95/index.d.ts#L109

RauliL commented:
Then I just cannot endorse this change in any way. Sorry.

katywings commented:
@raulil There is no difference between your idea from 
https://github.com/redom/redom/pull/160#issuecomment-529080308
and 
LINE 109
https://github.com/lufrai/redom/blob/bb805774d5940f06c864634c48686116a6038c95/index.d.ts#L109
your idea is already implemented.

Line 109 and Line 110 are not the f-ing same. As I said I can remove line 110 with the sooo evil *Function*. 

So whats your problem really?

katywings commented:
So I rebased and removed the lines with the evil `Function`.

While rebasing I noticed an actual real problem in the types from the master branch (#170). We should look at that separately.

So I wonder @raulil, since you even accepted a PR like #166, that clearly was unideal to say the least, if you no will work together with me and finish my PR. I removed the `Function` lines as a sign of diplomacy, to show you that we both actually have the same goals and I really hope that we can put our different thinking regarding strictness aside and get this thing done.

katywings commented:
Updated type resolution:
![Screenshot_20190911_122918](https://user-images.githubusercontent.com/4012401/64690095-e5a94b80-d48f-11e9-8e94-bb8e5f7a5247.png)

Source of "typeRedomComponentFunction" as seen in the screenshot:
```typescript
const typeRedomComponentFunction = function<El extends HTMLElement = HTMLElement, F = () => void>(fn: F): { new(): { el: El }} {
    return fn as any
}
```
This type helper function can be used to decorate redom component constructor functions with redom el type informations. If @raulil | @pakastin are interested we can add this function to redom. The function js in runtime itself is just `function(fn) { return fn }`.

pakastin commented:
What's the status of this pull request, ready to merge? Is there some kind of braking changes?

katywings commented:
Asked you like 5 months ago to merge 😂 on slack.

To make it a bit easier I now included the changes from https://github.com/redom/redom/issues/170 so that can be closed after this merge.

Regarding breaking change: personally I consider 95% of all typescript changes potentially breaking because you never can know what the lib-user makes with your types... On the other hand I dont have much experience with community feedback regarding ts. My opinion: If you think there are a lot ts redom users: make it major, otherwise maybe minor :)
